ansible_base/authentication/utils/claims.py

```diff
@@ -352,58 +401,279 @@ def create_organizations_and_teams(results) -> None:
+            # Cache role definition
+            if (role_definition := self._rd_by_id(role_assignment)) is None:
+                role_definition = role_assignment.role_definition
+                self.role_definitions[role_definition.name] = role_definition
+
+            # Cache Role User Assignment
+            self._init_key(role_definition.name, content_type_id=role_assignment.content_type_id)
+
+            # object_id is TEXT db type
+            object_id = int(role_assignment.object_id) if role_assignment.object_id is not None else None
+            obj = role_assignment.content_object if object_id else None
+
+            self.cache[role_definition.name][role_assignment.content_type_id][object_id] = {'object': obj, 'status': self.STATUS_EXISTING}
+
+    def rd_by_name(self, role_name: str) -> Optional[CommonModel]:
+        """Returns RoleDefinition by its name. Caches it if requested for first time"""
+        from ansible_base.rbac.models import RoleDefinition
+
+        try:
+            if self.role_definitions.get(role_name) is None:
+                self.role_definitions[role_name] = RoleDefinition.objects.get(name=role_name)
+        except ObjectDoesNotExist:
+            logger.warning(_("Skipping role '{role_name}', because the role does not exist").format(role_name=role_name))
+            self.role_definitions[role_name] = False  # skips multiple db queries
+
+        return self.role_definitions.get(role_name)
+
+    def _rd_by_id(self, role_assignment: models.Model) -> Optional[CommonModel]:
+        """Tries to find cached role definition by id, saving SQL queries"""
+        for rd in self.role_definitions.values():
+            if rd.id == role_assignment.role_definition_id:
+                return rd
+        return None
+
+    def add(self, role_name: str, organization: Optional[AbstractOrganization] = None, team: Optional[AbstractTeam] = None) -> None:
+        """Marks role assignment's params and (optionally) associated object in the cache.
+        If role_user_assignment (a.k.a. permission) existed before, marks it to do nothing
+        """
+        content_type_id = self._get_content_type_id(organization, team)
+        object_id = self._get_object_id(organization, team)
+
+        self._init_key(role_name, content_type_id=content_type_id)
+
+        # If the permission haven't existed, mark it to ADD, otherwise noop
+        current_status = self.cache[role_name][content_type_id].get(object_id, {}).get('status')
+        if current_status in [self.STATUS_EXISTING, self.STATUS_NOOP]:
+            self.cache[role_name][content_type_id][object_id] = {'object': organization or team, 'status': self.STATUS_NOOP}
+        elif current_status is None:
+            self.cache[role_name][content_type_id][object_id] = {'object': organization or team, 'status': self.STATUS_ADD}
+
+    def remove(self, role_name: str, organization: Optional[AbstractOrganization] = None, team: Optional[AbstractTeam] = None) -> None:
```

Comment:
Add and remove are almost identical and we have a couple places where we are doing:
```
if has_role:
    add
else:
    remove
```

Should we combine these two functions into one and have it take "has role" as a param and do the if inside of it?